Content.Shared/Alert/AlertsSystem.cs

```diff
@@ -0,0 +1,184 @@
+            if (alertsComponent.Alerts.TryGetValue(alert.AlertKey, out var alertStateCallback) &&
+                alertStateCallback.Type == alertType &&
+                alertStateCallback.Severity == severity &&
+                alertStateCallback.Cooldown == cooldown)
+            {
+                return;
+            }
+
+            // In the case we're changing the alert type but not the category, we need to remove it first.
+            alertsComponent.Alerts.Remove(alert.AlertKey);
+
+            alertsComponent.Alerts[alert.AlertKey] = new AlertState
+                { Cooldown = cooldown, Severity = severity, Type = alertType };
+
+            AfterShowAlert(alertsComponent);
+
+            alertsComponent.Dirty();
+        }
+        else
+        {
+            Logger.ErrorS("alert", "Unable to show alert {0}, please ensure this alertType has" +
+                                   " a corresponding YML alert prototype",
+                alertType);
+        }
+    }
+
+    /// <summary>
+    /// Clear the alert with the given category, if one is currently showing.
+    /// </summary>
+    public void ClearAlertCategory(EntityUid euid, AlertCategory category)
+    {
+        if(!EntityManager.TryGetComponent(euid, out AlertsComponent alertsComponent))
+            return;
+
+        var key = AlertKey.ForCategory(category);
+        if (!alertsComponent.Alerts.Remove(key))
+        {
+            return;
+        }
+
+        AfterClearAlert(alertsComponent);
+
+        alertsComponent.Dirty();
+    }
+
+    /// <summary>
+    /// Clear the alert of the given type if it is currently showing.
+    /// </summary>
+    public void ClearAlert(EntityUid euid, AlertType alertType)
+    {
+        if (!EntityManager.TryGetComponent(euid, out AlertsComponent alertsComponent))
+            return;
+
+        if (TryGet(alertType, out var alert))
+        {
+            if (!alertsComponent.Alerts.Remove(alert.AlertKey))
+            {
+                return;
+            }
+
+            AfterClearAlert(alertsComponent);
+
+            alertsComponent.Dirty();
+        }
+        else
+        {
+            Logger.ErrorS("alert", "unable to clear alert, unknown alertType {0}", alertType);
+        }
+    }
+
+    /// <summary>
+    /// Invoked after showing an alert prior to dirtying the component
+    /// </summary>
+    /// <param name="alertsComponent"></param>
+    protected virtual void AfterShowAlert(AlertsComponent alertsComponent) { }
+
+    /// <summary>
+    /// Invoked after clearing an alert prior to dirtying the component
+    /// </summary>
+    /// <param name="alertsComponent"></param>
+    protected virtual void AfterClearAlert(AlertsComponent alertsComponent) { }
+
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        LoadAlertPrototypes();
+    }
+
+    public void LoadAlertPrototypes()
+    {
+        foreach (var alert in _prototypeManager.EnumeratePrototypes<AlertPrototype>())
```

Comment:
I guess in order to support yaml reloading, this should be called whenever prototypes are reloaded (with a `_typeToAlert.Clear()`)

Reply:
ok